pkg/virt-controller/network/vmcontroller_test.go

```diff
@@ -0,0 +1,328 @@
+
+	networkv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+
+	"kubevirt.io/client-go/generated/kubevirt/clientset/versioned/fake"
+
+	v1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/libvmi"
+	"kubevirt.io/kubevirt/pkg/network/vmispec"
+	"kubevirt.io/kubevirt/pkg/virt-controller/network"
+)
+
+var _ = Describe("VM Network Controller", func() {
+	emptyVM := &v1.VirtualMachine{Spec: v1.VirtualMachineSpec{Template: &v1.VirtualMachineInstanceTemplateSpec{}}}
+
+	It("sync does nothing when the hotplug FG is unset", func() {
+		c := network.NewVMNetController(fake.NewSimpleClientset(), stubClusterConfig{}, stubPodGetter{})
+		originalVM := emptyVM.DeepCopy()
+		Expect(c.Sync(emptyVM, libvmi.New())).To(Equal(originalVM))
+	})
+
+	DescribeTable("sync does nothing when", func(vm *v1.VirtualMachine, vmi *v1.VirtualMachineInstance, podGetter stubPodGetter) {
+		c := network.NewVMNetController(fake.NewSimpleClientset(), stubClusterConfig{netHotplugEnabled: true}, podGetter)
+		originalVM := emptyVM.DeepCopy()
+		Expect(c.Sync(vm, vmi)).To(Equal(originalVM))
+	},
+		Entry("the VM is not running (there is no VMI)", emptyVM, nil, stubPodGetter{}),
+		Entry(
+			"the VMI is marked for deletion",
+			emptyVM,
+			&v1.VirtualMachineInstance{ObjectMeta: k8smetav1.ObjectMeta{DeletionTimestamp: &k8smetav1.Time{}}},
+			stubPodGetter{},
+		),
+		Entry("the VM & VMI have no interfaces and no pod is found", emptyVM, libvmi.New(), stubPodGetter{}),
+	)
+
+	It("sync fails when pod fetching returns an error", func() {
+		c := network.NewVMNetController(
+			fake.NewSimpleClientset(),
+			stubClusterConfig{netHotplugEnabled: true},
+			stubPodGetter{err: errors.New("test")},
+		)
+		originalVM := emptyVM.DeepCopy()
+		updatedVM, err := c.Sync(emptyVM, libvmi.New())
+		Expect(err).To(MatchError(isSyncErrorType, "syncError"))
+		Expect(updatedVM).To(Equal(originalVM))
+	})
+
+	It("sync fails when pod patch returns an error", func() {
+		clientset := fake.NewSimpleClientset()
+		c := network.NewVMNetController(
+			clientset,
+			stubClusterConfig{netHotplugEnabled: true},
+			stubPodGetter{pod: &k8sv1.Pod{}},
+		)
+
+		// Setup `Patch` to fail.
+		injectedPatchError := errors.New("test patch error")
+		clientset.Fake.PrependReactor("patch", "virtualmachineinstances", func(action testing.Action) (handled bool, obj k8sruntime.Object, err error) {
+			return true, nil, injectedPatchError
+		})
+
+		vmi := libvmi.New(
+			libvmi.WithInterface(libvmi.InterfaceDeviceWithMasqueradeBinding()),
+			libvmi.WithNetwork(v1.DefaultPodNetwork()),
+		)
+		vm := libvmi.NewVirtualMachine(vmi.DeepCopy())
+
+		vm = plugNetworkInterface(vm, "foonet")
+
+		// Simulate the existence of the VMI on the server (to allow the Sync to patch it).
+		_, err := clientset.KubevirtV1().VirtualMachineInstances(vmi.Namespace).Create(context.Background(), vmi, k8smetav1.CreateOptions{})
+		Expect(err).NotTo(HaveOccurred())
```

Comment:
nit: The vmi could be passed to 
```go
clientset := fake.NewSimpleClientset(vmi)
```

and make this create operation redundant.
Same bellow.

Reply:
Nice, but it will require to create the VMI before creating the net-controller which is odd.